cpp/src/arrow/sparse_tensor.cc

```diff
@@ -301,6 +301,127 @@ class SparseTensorConverter<TYPE, SparseCSRIndex>
   inline Status CheckMaximumValue(const uint64_t) const { return Status::OK(); }
 };
 
+// ----------------------------------------------------------------------
+// SparseTensorConverter for SparseCSCIndex
+
+template <typename TYPE>
+class SparseTensorConverter<TYPE, SparseCSCIndex>
+    : private SparseTensorConverterBase<TYPE> {
+ public:
+  using BaseClass = SparseTensorConverterBase<TYPE>;
+  using NumericTensorType = typename BaseClass::NumericTensorType;
+  using value_type = typename BaseClass::value_type;
+
+  SparseTensorConverter(const NumericTensorType& tensor,
+                        const std::shared_ptr<DataType>& index_value_type,
+                        MemoryPool* pool)
+      : BaseClass(tensor, index_value_type, pool) {}
+
+  template <typename IndexValueType>
+  Status Convert() {
+    using c_index_value_type = typename IndexValueType::c_type;
+    RETURN_NOT_OK(CheckMaximumValue(std::numeric_limits<c_index_value_type>::max()));
+    const int64_t indices_elsize = sizeof(c_index_value_type);
+
+    const int64_t ndim = tensor_.ndim();
+    if (ndim > 2) {
+      // LCOV_EXCL_START: The following invalid causes program failure.
+      return Status::Invalid("Invalid tensor dimension");
+      // LCOV_EXCL_STOP
+    }
+
+    const int64_t nr = tensor_.shape()[0];
+    const int64_t nc = tensor_.shape()[1];
+    int64_t nonzero_count = -1;
+    RETURN_NOT_OK(tensor_.CountNonZero(&nonzero_count));
+
+    std::shared_ptr<Buffer> indptr_buffer;
+    std::shared_ptr<Buffer> indices_buffer;
+
+    std::shared_ptr<Buffer> values_buffer;
+    RETURN_NOT_OK(
+        AllocateBuffer(pool_, sizeof(value_type) * nonzero_count, &values_buffer));
+    value_type* values = reinterpret_cast<value_type*>(values_buffer->mutable_data());
+
+    if (ndim <= 1) {
+      return Status::NotImplemented("TODO for ndim <= 1");
```

Comment:
Why TODO? Does it make sense to use CSC for a 1-dimensional tensor?

Reply:
Perhaps this we could implement this as Nx1 dimensional CSC?

Reply:
We can support creating CSC and CSR matrices from 0-dimensional or 1-dimensional dense tensors.

Reply:
How does compression work in that case?

Reply:
Perhaps just limiting to `ndim == 2` is best? CSC and CSR are by definition 2D types, right?

Reply:
SciPy's `csr_matrix` and `csc_matrix` can be created from NumPy's 0-dimensional and 1-dimensional arrays.
I think it is better that we support creating a `SparseCSRMatrix` and a `SparseCSCMatrix` from 0-dimensional and 1-dimensional `Tensor`s.

I will do it in https://issues.apache.org/jira/browse/ARROW-7421